🏗️ Overhaul frontend

Benjamin-Piper opened the pull request:
# In a nutshell

This PR overhauls the frontend in order to create a better developer
experience.

- **Faster**: move from [webpack] to [vite].
- **Clearer and more organised**: new file structure and some renamed
  files.

[webpack]: https://webpack.js.org/
[vite]: https://vite.dev/

**Proof that it works!** https://benjamin-piper.github.io/eberban/

---

## `/dictionary/`

- 🟩 Add `package.json` that only includes `npm run ids` script.
- ➡️ Move `generate_word_ids` to here.

**Reason:** Dictionary editors need only visit this directory.

## `/web/`

Everything frontend-code related is found under this directory.

- ➡️ Move `images/` to here.
- 🟥🟩 Delete old `package.json` and `package-lock.json` and create new
  ones here.
- 🟩 Created `src/` here and consolidated all of `node/` and `web/`
  here.

**Reason:** Website editors need only visit this directory.

### `/web/src/`

- ➡️ Move `index.html` and `index.css` to here.
- 🟥🟩 Delete `webpack.config.js` and create `vite.config.js` here.
- ➡️ Move webpages and related code to their own directories:
   -  `/web/src/dictionary/ `
      - 🟨 Rename `src/dictionary.js` to `compare_words.js` and move it
        to here.
   - `/web/src/root-generator/`
      - 🟨 Rename `root_generation/index.js` to `random_root.js` and
        move it to here.
   - `/web/src/textual-parser/` (previously `parser_text_dynamic.html`)
   - `/web/src/visual-parser/` (previously `parser_box_glosser.html`)
- ➡️ Move `grammar/` to here.
- 🟩➡️ Create `legacy/` and move all legacy files to it.
- 🟨 Update many files to move from [CommonJS syntax] to
  [ECMAScript Module syntax].
- ➡️ Move `scripts/` to here.

[CommonJS syntax]: https://nodejs.org/api/modules.html
[ECMAScript Module syntax]: https://nodejs.org/api/esm.html

**Reasons:**

- **More organised:** I found it cleaner to have the root of the vite
  project in `/web/src/`.
- **Faster:** Vite is fast 😎. And it uses JavaScript modules.
- **More organised:** Webpages and related code in their own directories
  makes it easier to focus on specific changes.
- **Clearer:** Used more descriptive names.

#### `/web/src/scripts`

- ~~🟥 Remove `all_non_init_pairs.js`.~~
- ➡️Move all parsing code to `/web/src/scripts/parsing/`.
- ➡️Move all peg code to `/web/src/scripts/peg/`.
   - 🟨 Rename `build_peg` to `pegjs_to_js`.

**Reasons:**

- **More organised:** ~~`all_non_init_pairs.js` wasn't being used AFAIK.~~
  Also, I found it cleaner to group related scripts into directories.
- **Clearer:** Renamed `build_peg` to clearly say what it's doing.

> [!NOTE]
> Using our own peg scripts like this is not ideal. Valuing simplicity
> and speed, we should be teaching the vite config how to build our
> `eberban.peg` file. This is out of scope for this PR though.

## Other files updated accordingly

- 🟨 `.gitignore`.
- 🟨 `README`.
- 🟨 `.github/workflows/gh-pages.yml` (this also includes upgrading
  `actions/checkout`).

Benjamin-Piper committed:
🏗️ Overhaul frontend

# In a nutshell

This PR overhauls the frontend in order to create a better developer
experience.

- **Faster**: move from [webpack] to [vite].
- **Clearer and more organised**: new file structure and some renamed
  files.

[webpack]: https://webpack.js.org/
[vite]: https://vite.dev/

**Proof that it works!** https://benjamin-piper.github.io/eberban/

---

## `/dictionary/`

- 🟩 Add `package.json` that only includes `npm run ids` script.
- ➡️ Move `generate_word_ids` to here.

**Reason:** Dictionary editors need only visit this directory.

## `/web/`

Everything frontend-code related is found under this directory.

- ➡️ Move `images/` to here.
- 🟥🟩 Delete old `package.json` and `package-lock.json` and create new
  ones here.
- 🟩 Created `src/` here and consolidated all of `node/` and `web/`
  here.

**Reason:** Website editors need only visit this directory.

### `/web/src/`

- ➡️ Move `index.html` and `index.css` to here.
- 🟥🟩 Delete `webpack.config.js` and create `vite.config.js` here.
- ➡️ Move webpages and related code to their own directories:
   -  `/web/src/dictionary/ `
      - 🟨 Rename `src/dictionary.js` to `compare_words.js` and move it
        to here.
   - `/web/src/root-generator/`
      - 🟨 Rename `root_generation/index.js` to `random_root.js` and
        move it to here.
   - `/web/src/textual-parser/` (previously `parser_text_dynamic.html`)
   - `/web/src/visual-parser/` (previously `parser_box_glosser.html`)
- ➡️ Move `grammar/` to here.
- 🟩➡️ Create `legacy/` and move all legacy files to it.
- 🟨 Update many files to move from [CommonJS syntax] to
  [ECMAScript Module syntax].
- ➡️ Move `scripts/` to here.

[CommonJS syntax]: https://nodejs.org/api/modules.html
[ECMAScript Module syntax]: https://nodejs.org/api/esm.html

**Reasons:**

- **More organised:** I found it cleaner to have the root of the vite
  project in `/web/src/`.
- **Faster:** Vite is fast 😎. And it uses JavaScript modules.
- **More organised:** Webpages and related code in their own directories
  makes it easier to focus on specific changes.
- **Clearer:** Used more descriptive names.

#### `/web/src/scripts`

- 🟥 Remove `all_non_init_pairs.js`.
- ➡️Move all parsing code to `/web/src/scripts/parsing/`.
- ➡️Move all peg code to `/web/src/scripts/peg/`.
   - 🟨 Rename `build_peg` to `pegjs_to_js`.

**Reasons:**

- **More organised:** `all_non_init_pairs.js` wasn't being used AFAIK.
  Also, I found it cleaner to group related scripts into directories.
- **Clearer:** Renamed `build_peg` to clearly say what it's doing.

> [!NOTE]
> Using our own peg scripts like this is not ideal. Valuing simplicity
> and speed, we should be teaching the vite config how to build our
> `eberban.peg` file. This is out of scope for this PR though.

## Other files updated accordingly

- 🟨 `.gitignore`.
- 🟨 `README`.
- 🟨 `.github/workflows/gh-pages.yml` (this also includes upgrading
  `actions/checkout`).

mia-entropy commented on README.md:
```diff
@@ -38,48 +38,58 @@ If you would like to get involved, please also consider joining our
+The dev server will update this local website after every change you make.
 
-[`miniserve`]: https://lib.rs/crates/miniserve
+[`mdbook`]: https://lib.rs/crates/mdbook
```
Now unused. Would be great to put back intro that books are made with mdbook (which justified the link).

Benjamin-Piper committed:
🔨 Re-add scripts/all_non_init_pairs.js

I initially removed this because it was unused. But after talking with
@mia-entropy, they told me its use and that they'd prefer to keep it
just in case.

> all_non_init_pairs.js was a piece of code that allowed to generate
> fromthe parser the list of medial pairs to ensure the list is correct
> and the rules are well written.
>
> i'd prefer to keep it just in case

Benjamin-Piper committed:
📝 Re-add to README that books are made using `mdbook`

Makes use of the [`mdbook`] link.

Benjamin-Piper committed:
🐛 Fix (via rewrite) dictionary searching

I'm not sure why the searching functionality broke after the
"🏗️ overhaul frontend"

I took the time to rewrite and simplify this and remove jQuery. The
result is something much simpler, I think.
